### PR TITLE
[BUD-126]Redirection of user based on role

### PR DESCRIPTION
**Pull request by michaelmoney**
# [BUD-126](https://digitalrig.atlassian.net/browse/BUD-126) Redirection of user based on role

## Description
Generic component that redirects user to correct path based on authorisation state and user role type (newbie/buddy) and support protected routes.

## Checklist (optional)

- [x] Redirects user based on the role (newbie / buddy) after successful login
- [x] Redirects user to default path if user provides non existing route
- [x] Protect routes from being accessible if user is not logged in and redirects to `/login`

### CODE REVIEW

File: src/components/RouteRedirect/RouteRedirect.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      return (
+        <Redirect
+          to={{
+            pathname: location.pathname,
+          }}
+        />
+      );
+    }
+  }
+  return <></>;
```
Comment (lucianojss) [inline src/components/RouteRedirect/RouteRedirect.tsx]: What happens if the code hit this case?

File: src/utils/user-helper.ts [hunk 2]
```diff
@@ -0,0 +1,2 @@
+export const isBuddy = (role: string) => role === 'BUDDY';
```
Comment (lucianojss) [inline src/utils/user-helper.ts]: Isn't possible to pick this roles from the generated schema ENUM?

File: src/components/ProtectedRoute/ProtectedRoute.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import { Route, Redirect } from 'react-router-dom';
+import userService from '../../utils/user-service';
+import { ROUTES } from '../../shared/routes';
+
+const ProtectedRoute: React.FC<any> = ({ component: Component, ...rest }) => {
```
Comment (lucianojss) [inline src/components/ProtectedRoute/ProtectedRoute.tsx]: Can't we fix this `any`?

**Comment by lucianojss** [comment]
Good job @michaelmoney 👍 
I would like to see this implemented with react hooks context instead of using a service, because it can allow us to add extra information to the user state, for example some client settings etc. 

Another point I would like to point out is, even if we add cookies with fake data, the app is going to try to it the api and will get 401 error but it will kept logged in (correct me if I am wrong). we should add a case that if we get 401 error we should logout and redirect to base route.

File: src/components/AppWrapper/__tests__/__snapshots__/AppWrapper.test.tsx.snap [hunk 4]
```diff
@@ -6,5 +6,127 @@ exports[`Component - AppWrapper renders correctly 1`] = `
   component="main"
   data-testid="root"
   maxWidth="md"
-/>
+>
+  <RouteRedirect
+    history={
```
[the remark this answers is not in the document]
Comment (michaelmoney) [reply src/components/AppWrapper/__tests__/__snapshots__/AppWrapper.test.tsx.snap]: Well.... **ProtectedRoute** component *is* mocked 😀 Look at `src/components/AppWrapper/__tests__/AppWrapper.test.tsx`

```
jest.mock('../../ProtectedRoute/ProtectedRoute', () => 'ProtectedRoute');
jest.mock('../../RouteRedirect/RouteRedirect', () => 'RouteRedirect');
```

I guess snapshots takes props from `withRouter`, but don't know how to mock it. If you have any ideas, pls let me know.

File: src/components/AppWrapper/__tests__/__snapshots__/AppWrapper.test.tsx.snap [hunk 5]
```diff
@@ -6,5 +6,127 @@ exports[`Component - AppWrapper renders correctly 1`] = `
...
+  />
+  <ProtectedRoute
+    component={[Function]}
+    exact={true}
+    path="/newbie/tasks/:taskId"
+  />
+  <RouteRedirect
+    history={
+      Object {
+        "action": "POP",
```
[the remark this answers is not in the document]
Comment (michaelmoney) [reply src/components/AppWrapper/__tests__/__snapshots__/AppWrapper.test.tsx.snap]: ☝️

File: src/components/ProtectedRoute/ProtectedRoute.tsx (hunk 3 above)
Comment (michaelmoney) [reply src/components/ProtectedRoute/ProtectedRoute.tsx]: Fixed in: 490058b8c708b9b705a09044e59b350c11733afb 🍺

File: src/components/AppWrapper/AppWrapper.tsx [hunk 6]
```diff
@@ -12,6 +11,9 @@ import NewbieSelect from '../NewbieSelect';
 import ContactDetails from '../ContactDetails';
 import AddTask from '../AddTask';
 import ErrorPage from '../ErrorPage';
+import ProtectedRoute from '../ProtectedRoute/ProtectedRoute';
+import Logout from '../Logout/Logout';
+import RouteRedirect from '../RouteRedirect/RouteRedirect';
```
[the remark this answers is not in the document]
Comment (michaelmoney) [reply src/components/AppWrapper/AppWrapper.tsx]: Fixed in: 490058b8c708b9b705a09044e59b350c11733afb 👍

File: src/components/AppWrapper/AppWrapper.tsx [hunk 7]
```diff
@@ -12,6 +11,9 @@ import NewbieSelect from '../NewbieSelect';
 import ContactDetails from '../ContactDetails';
 import AddTask from '../AddTask';
 import ErrorPage from '../ErrorPage';
+import ProtectedRoute from '../ProtectedRoute/ProtectedRoute';
```
[the remark this answers is not in the document]
Comment (michaelmoney) [reply src/components/AppWrapper/AppWrapper.tsx]: Fixed in: 490058b8c708b9b705a09044e59b350c11733afb 👍

**Comment by michaelmoney** [comment]
@lucianojss 

> Good job @michaelmoney 👍
> I would like to see this implemented with react hooks context instead of using a service, because it can allow us to add extra information to the user state in future, for example some client settings etc.

If it's gonna fit us better let's do it. Service was really simple way to handle `id` and `role` variables which remains unchanged during user session. 

> Another point I would like to point out is, even if we add cookies with fake data, the app is going to try to it the api and will get 401 error but it will kept logged in (correct me if I am wrong). we should add a case that if we get 401 error we should logout and redirect to base route.

Yes, that's true and i'm aware of that. From security perspective that's not an issue, since user will be unauthorised and error will be thrown. This kind of error should be globally handled and logout user (remove cookies) and redriect to the `/login` screen.

File: src/components/Login/Login.tsx [hunk 8]
```diff
@@ -45,7 +45,8 @@ const Login = () => {
   const [loginMutation, { loading }] = useMutation<AuthData>(LOGIN_MUTATION, {
     onCompleted: ({ login }) => {
       auth.setToken(login.token);
-      history.push(ROUTES.BUDDY_SELECT_NEWBIE);
+      auth.setUser({ id: login.userId, role: login.role });
```
Comment (Tonik1204) [inline src/components/Login/Login.tsx]: `auth.setToken` and `auth. setUser ` should be changed to use `AuthContext`. We have create separate folder `/src/context` with Auth part

File: src/components/Logout/Logout.tsx [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
```
Comment (Tonik1204) [inline src/components/Logout/Logout.tsx]: I belive we don't need logout component

File: src/components/ProtectedRoute/ProtectedRoute.tsx [hunk 10]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
```
Comment (Tonik1204) [inline src/components/ProtectedRoute/ProtectedRoute.tsx]: I belive we don't need this component

File: src/components/RouteRedirect/RouteRedirect.tsx [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+};
+
+export default withRouter(RouteRedirect);
```
Comment (Tonik1204) [inline src/components/RouteRedirect/RouteRedirect.tsx]: I belive we don't need this component

File: src/shared/routes.ts [hunk 12]
```diff
@@ -1,6 +1,7 @@
 enum ROUTES {
   BASE = '/',
   LOGIN = '/login',
+  LOGOUT = '/logout',
```
Comment (Tonik1204) [inline src/shared/routes.ts]: I belive we don't need this route

File: src/utils/cookie-service.ts [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+  deleteUser() {
+    Cookies.remove(AUTH_KEYS.USER);
+  }
+  deleteToken() {
+    Cookies.remove(process.env.REACT_APP_AUTH_TOKEN || AUTH_KEYS.TOKEN);
+  }
+}
+
+const cookieService = new CookieService();
+
```
Comment (Tonik1204) [inline src/utils/cookie-service.ts]: It has to be in AuthContext

File: src/utils/user-helper.ts (hunk 2 above)
Comment (Tonik1204) [reply src/utils/user-helper.ts]: Then we need to add similar schema-types from here `buddy-app/server/src/generated/schema-types.ts` file to `src/types` folder for example

File: src/utils/user-service.ts [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+  logout(): void {
+    cookieService.deleteToken();
+    cookieService.deleteUser();
+  }
+}
+
+const userService = new UserService();
+
+export default userService;
```
Comment (Tonik1204) [inline src/utils/user-service.ts]: It has to be in AuthContext

**Comment by michaelmoney** [comment]
> @michaelmoney There's a big job, for sure. But still I see the way how it can be improved:
> First I would like to update routs logic - to use `Switch` component with coditional `Routes` and `Redirect` same as in `AppWrapper.tsx.zip` below:
> [AppWrapper.tsx.zip](https://github.com/BuildIt-Poland/buddy-app/files/3816565/AppWrapper.tsx.zip)

I can download file, but can't unzip it.. 🤔 I guess sth is wrong.

> Thus we can get rig of RouteRedirect, ProtectedRoute and Logut components.
> Then we just need to update AuthContext whenever user is loggedIn/Out and use it in AppWrapper.
> And next would be really cool to add AuthContext instead of user/cookies services.
> If you need a help or more explanations - I'm always glad to help you ;)

Looks like you are proposing totally different approach. I think it's going to be hard to change this PR into new implementation (lots of tests and components to be removed etc) based on AuthContext and Switch. I think it's just better to start from the scratch. That's being said, maybe you could submit a PR and I will decline mine?

**Comment by Tonik1204** [comment]
> > @michaelmoney There's a big job, for sure. But still I see the way how it can be improved:
> > First I would like to update routs logic - to use `Switch` component with coditional `Routes` and `Redirect` same as in `AppWrapper.tsx.zip` below:
> > [AppWrapper.tsx.zip](https://github.com/BuildIt-Poland/buddy-app/files/3816565/AppWrapper.tsx.zip)
> 
> I can download file, but can't unzip it.. 🤔 I guess sth is wrong.
> 
> > Thus we can get rig of RouteRedirect, ProtectedRoute and Logut components.
> > Then we just need to update AuthContext whenever user is loggedIn/Out and use it in AppWrapper.
> > And next would be really cool to add AuthContext instead of user/cookies services.
> > If you need a help or more explanations - I'm always glad to help you ;)
> 
> Looks like you are proposing totally different approach. I think it's going to be hard to change this PR into new implementation (lots of tests and components to be removed etc) based on AuthContext and Switch. I think it's just better to start from the scratch. That's being said, maybe you could submit a PR and I will decline mine?

**Comment by Tonik1204** [comment]
Sorry, mistakenly closed it. Yeah, that's right, I should prepare another pr

**Comment by michaelmoney** [comment]
Closing this one ⛔️. As discussed with @Tonik1204 - there will be new implementation for **[[BUD-126]Redirection of user based on role](https://digitalrig.atlassian.net/browse/BUD-126)**.

**Pull request closed without merging**